File: carpark_queueing_web.py

#Import Functions Required
import random
import itertools
import time
import sys

arrivalrate = int(sys.argv[1])
servicetime = int(sys.argv[2])
spaces = int(sys.argv[3])

#Define Functions
def percentage(percent,list):
    for index, number in enumerate(list):
        if number >= percent:
            return index

def modelrun(arrivalrate,servicetime,spaces):
    arrival = 0
    count_arrivals = 0
    count_carsparked_q = [0] * int(arrivalrate * int(servicetime/600)+200)
    count_carsqueued = [0] * int(arrivalrate * int(servicetime/600)+200)
    carsparked_q = []
    carsqueued = 0
    queue = 0
    queuetime = 0
    queuetest = 0
    hours = 0
    cyclecount = int(10000)

    #Generate Arrivals
    
    for i in range (1,cyclecount * 3600 ):
        if i  > 3600*100 and i%36000 ==0:
            if queuetest == round(carsqueued/count_arrivals,5) or i == 3000*3600:
                hours = i / 3600
                break
            else:
                queuetest = round(carsqueued/count_arrivals,5)
                
        # Count current carpark utilisation
        count_carsparked_q[max(len(carsparked_q)-1,0)] += 1
        count_carsqueued[queue] += 1
        queuetime += queue

        # Reduce parked cars time remaining by passing time
        if len(carsparked_q) > 0:
            carsparked_q = [item - 1 for item in carsparked_q]

            # move finished cars out and queued cars in    
            if carsparked_q[0] == 0:
                del carsparked_q[0]
                if queue > 0:
                    carsparked_q.append(servicetime)
                    queue -= 1

        # Check if new car arrived and at to carpark
        arrival = random.randint(1,3600 * precision)
        if arrival <= arrivalrate * precision:
            count_arrivals += 1
            if len(carsparked_q) - 1 < spaces:
                carsparked_q.append(servicetime)
            if len(carsparked_q) - 1 == spaces:
                queue +=1
                carsqueued +=1
            if len(carsparked_q) - 1 > spaces:
                print("more cars than spaces somehow")

    cyclecount = hours

    #Convert Counts to Percentiles
    count_carsqueued = list(itertools.accumulate(count_carsqueued))
    count_carsparked_q = list(itertools.accumulate(count_carsparked_q))
    if cyclecount > 0:
        count_carsqueued = [round(100 * item / (cyclecount * 3600 * precision),2) for item in count_carsqueued]
        count_carsparked_q = [round(100 * item / (cyclecount * 3600 * precision),2) for item in count_carsparked_q]

    for i in range(0,100):
        for index, number in enumerate(count_carsqueued):
                if number >= i:
                    count_carsqueued.append(index)
                    break
    for i in range(0,100):
        for index, number in enumerate(count_carsparked_q):
                if number >= i:
                    if number <= count_carsqueued[0]:
                        count_carsparked_q.append(index)
                        break
                        # Simple Data Outputs
    end_time = time.time()
    elapsed_time = end_time - start_time
    print("Model completed in ", int(round(elapsed_time,0)), " seconds", sep='')
    if hours == 3000:
        print("0.00001 stability not found. ",hours,"hours of survey data generated.")
    else:
        print("Stable solution found after",hours,"hours of survey data")
    if cyclecount >0:
        print("Random Arrivals = ", round(count_arrivals / cyclecount / precision,1))
        print("Service Time = ", servicetime)
        print("Linear Spaces of Demand = ", round(count_arrivals / cyclecount / precision * servicetime / 3600,2),"spaces")
    
    print("Cars Queued = ", round(carsqueued * 100 / count_arrivals,2),"%", sep='')
    print("Average Queue time per Arrival = ", round(queuetime / count_arrivals),"seconds")
    if carsqueued > 0:    
        print("Average Queue time per Queued Vehicle = ", round(queuetime / carsqueued),"seconds")
        

    print("Operating Combined Percentiles")
    for value in percentiles:
        print(value,"th percentile is ",percentage(value,count_carsparked_q)," parked and ", percentage(value,count_carsqueued)," Queued", sep='')

# print("How many hours to run for?")
while True:
    try:
        # Prompt the user to enter an integer
        cyclecount = 100000 # int(input("Minimum 1000: "))
        if cyclecount > 999:
            break  # Exit the loop if input is valid
    except ValueError:
        # Handle the error if input is not a valid integer
        print("That's not a valid number. Please try again.")


#print("Cycles per second for precision?")
precision = 1 #int(input("Enter number above 0: "))

#Intialise All Variables
arrival = 0
count_arrivals = 0
utilisation = []
count_carsparked_q = [0] * int(arrivalrate * int(servicetime/600)+2000)
count_carsqueued = [0] * int(arrivalrate * int(servicetime/600)+2000)
carsparked_q = []
carsqueued = 0
queue = 0
last_q = 0
percentiles = [10,20,30,40,50,60,70,80,90,95,98,99]
queuedgraph = []
parkedgraph = []
queuetime = 0
queuetest = 0
hours = 0

# os.system('clear')
start_time = time.time()  

modelrun(arrivalrate,servicetime,spaces)

File: carpark_queueing_web.js
var precision = 1;
var HOURS_IN_SECONDS = 3600 * precision;

function percentageOfTime(percent, list) {
    let cumulativeSum = 0;
    const totalTime = list.reduce((sum, value) => sum + value, 0);
    const targetSum = (totalTime * percent) / 100;

    for (let index = 0; index < list.length; index++) {
        cumulativeSum += list[index];
        if (cumulativeSum >= targetSum) {
            return index;
        }
    }
    return 0;
}

function calculatePercentiles(countCarsParked, countCarsQueued) {
    const percentiles = [10, 20, 30, 40, 50, 60, 70, 80, 90, 95, 98, 99];
    const results = new Map();

    for (const p of percentiles) {
        results.set(p, {
            parked: percentageOfTime(p, countCarsParked),
            queued: percentageOfTime(p, countCarsQueued)
        });
    }

    return Object.fromEntries(results);
}

function runSimulation(inp) {
    console.dir(inp, { depth: 2 });
    const startTime = performance.now();

    const state = {
        carsParked: [],
        carsQueued: 0,
        countArrivals: 0,
        countCarsParked: new Array(inp.spaces + 1).fill(0),
        countCarsQueued: new Array(Math.ceil(inp.arrivalRate * inp.serviceTime)).fill(0),
        queue: 0,
        queueTest: 0,
        queueTime: 0
    };

    let hours = 0;
    for (let i = 1; i <= inp.cycles * HOURS_IN_SECONDS; i++) {
        if (i > 3600*500*precision && i % 36000*precision === 0) {
            hours = i / HOURS_IN_SECONDS;
            const currentQueueRatio = state.carsQueued / state.countArrivals;
            if (Math.abs(state.queueTest - currentQueueRatio) <= 1e-6) {
                break;
            } else {
                state.queueTest = currentQueueRatio;
            }
        }

        // Check if a new car arrived and add to the car park
        const arrival = Math.floor(Math.random() * (HOURS_IN_SECONDS * inp.precision)) + 1;
        if (arrival <= inp.arrivalRate * inp.precision) {
            state.countArrivals++;
            if (state.carsParked.length < inp.spaces) {
                state.carsParked.push(inp.serviceTime*inp.precision);
            } else {
                state.queue++;
                state.carsQueued++;
            }
        }

        // Count current car park utilization
        const index = Math.max(state.carsParked.length, 0);
        state.countCarsParked[index]++;
        state.countCarsQueued[state.queue]++;
        state.queueTime += state.queue;

        // Reduce parked cars' time remaining
        if (state.carsParked.length >= 1) {
            for (let j = 0; j < state.carsParked.length; j++) {
                state.carsParked[j]--;
            }

            // Move finished cars out and queued cars in
            if (state.carsParked[0] === 0) {
                state.carsParked.shift();
                if (state.queue > 0) {
                    state.carsParked.push(inp.serviceTime*inp.precision);
                    state.queue--;
                }
            }
        }
    }
    const elapsedMs = performance.now() - startTime;
    return {
        elapsedTime: `${Math.round(elapsedMs)}ms`,
        hours,
        PercentCarsQueued: ((state.carsQueued * 100) / state.countArrivals).toFixed(1),
        QueueTimePerQueuedVehicle: (state.queueTime / state.carsQueued / precision).toFixed(1),
        QueueTimePerArrival: (state.queueTime / state.countArrivals / precision).toFixed(1),
        theoreticalMinSpace: ((state.countArrivals / hours * inp.serviceTime) / HOURS_IN_SECONDS * precision).toFixed(2),
        modelDemandSpace: ((inp.arrivalRate * inp.serviceTime) / HOURS_IN_SECONDS * precision).toFixed(2),
        percentiles: calculatePercentiles(state.countCarsParked, state.countCarsQueued)
    };
}

// Format results as a readable string
export function formatResults(results) {
    if (!results || typeof results !== 'object') {
        return 'Error: Invalid simulation results.';
    }

    let output = `Queueing Results:\n`;
    output += `-----------------------------\n`;
    output += `Elapsed Time: ${results.elapsedTime || 'N/A'}\n`;
    output += `Simulated Hours: ${results.hours || 'N/A'}\n`;
    output += `Percentage of Cars Queued: ${results.PercentCarsQueued || 'N/A'}%\n`;
    output += `Avg. Queue Time per Queued Vehicle: ${results.QueueTimePerQueuedVehicle || 'N/A'} sec\n`;
    output += `Avg. Queue Time per Arrival: ${results.QueueTimePerArrival || 'N/A'} sec\n`;
    output += `Theoretical Min Spaces: ${results.theoreticalMinSpace || 'N/A'}\n`;
    output += `Model Demand Spaces: ${results.modelDemandSpace || 'N/A'}\n`;
    output += `-----------------------------\n`;
    output += `Percentiles (Parked | Queued):\n`;

    // ✅ Ensure percentiles are correctly formatted
    if (results.percentiles && typeof results.percentiles === 'object') {
        for (const [percentile, values] of Object.entries(results.percentiles)) {
            output += `${percentile}%: ${values.parked ?? 'N/A'} | ${values.queued ?? 'N/A'}\n`;
        }
    } else {
        output += `Percentiles data not available.\n`;
    }
    return output;
}

function getDecimalPlaces(num) {
    if (!num.toString().includes('.')) return 0; // No decimal places
    return num.toString().split('.')[1].length;  // Count decimal places
}

// Main function to simulate the process
export function main(a, b, c) {

    // // Convert inputs to numbers
    // const numA = Number(a);
    // const numB = Number(b);
    // const numC = Number(c);

    // // Find the max decimal places
    // const maxDecimalPlaces = Math.max(
    //     getDecimalPlaces(numA), 
    //     getDecimalPlaces(numB), 
    //     getDecimalPlaces(numC)
    // );

    // // Set precision as 10^maxDecimalPlaces
    // precision = Math.pow(10, maxDecimalPlaces);
    precision = 1;
    const inputs = {
        arrivalRate: Number(a),
        cycles: 10000,
        precision: precision,
        HOURS_IN_SECONDS: 3600*precision,
        serviceTime: Number(b),
        spaces: Number(c)
    };
    HOURS_IN_SECONDS = 3600 * precision;
    try {
        const results = runSimulation(inputs);
        return formatResults(results);
    } catch (error) {
        console.error('Error running simulation:', error);
    }
}

